plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/graphical/figures/SiriusLayoutHelper.java

```diff
@@ -87,14 +93,91 @@ public Point getReferencePosition(IFigure parent, Viewport viewport, IGraphicalE
+        return snapLocation;
+    }
+
+    /**
+     * Overridden to snap after shiftng the point.
+     */
+    public Point validatePosition(IFigure parent, Rectangle bounds) {
+        Rectangle theBounds = bounds.getCopy();
+        IFigure clobber = findFigureIn(parent, bounds);
+        if (clobber != null) {
+            theBounds.setLocation(updateClobberedPosition(clobber, null));
+
+            // Snap
+            Point location = theBounds.getLocation();
+            GraphicalHelper.logical2screen(location, this.containerEditPart);
+            EditPartQuery editPartQuery = new EditPartQuery(this.containerEditPart);
+            Point snapLocation = editPartQuery.getSnapLocation(new CreateRequest(), location);
+            GraphicalHelper.screen2logical(snapLocation, this.containerEditPart);
+
+            theBounds.setLocation(snapLocation);
+
+            return validatePosition(parent, theBounds);
+        }
+        return theBounds.getLocation();
     }
 
     @Override
     public Point updateClobberedPosition(IFigure clobbered, IFigure newlyAddedChild) {
-        Point clobberedPosition = super.updateClobberedPosition(clobbered, newlyAddedChild);
+        Point clobberedPosition = computeTranslatedPoint(determineReferencePoint(clobbered), clobbered, true);
+        return clobberedPosition;
+    }
+
+    /**
+     * Determines the figure reference point based on user preferences. This point can then be used to determine the
+     * arrangement of adjacent figures.
+     * 
+     * @param figure
+     *            the figure for which the reference point is determined
+     * @return the figure reference point
+     */
+    public Point determineReferencePoint(IFigure figure) {
+        int pref = DiagramUIPlugin.getPlugin().getPreferenceStore().getInt(SiriusDiagramUiPreferencesKeys.PREF_NEWLY_CREATED_ELEMENTS_LAYOUT.name());
+        Dimension size = figure.getSize().getCopy().union(figure.getPreferredSize());
+        Rectangle bounds = figure.getBounds().getCopy().setSize(size);
+        Point referencePoint = switch (pref) {
+        case SiriusLayoutDataManager.VERTICAL_ARRANGEMENT -> bounds.getBottomLeft().getCopy();
+        case SiriusLayoutDataManager.HORIZONTAL_ARRANGEMENT -> bounds.getTopRight().getCopy();
+        default -> bounds.getRight().getCopy();
+        };
+        return referencePoint;
+    }
+
+    /**
+     * Computes the translated point based on user preferences and a reference point. The translation is determined by
+     * user preferences and applied to the specified reference point.
+     * 
+     * @param referencePoint
+     *            the reference point to translate
+     * @param figure
+     *            the figure for which to compute the translated point
+     * @param shiftWithGridValue 
+     * @return the translated point based on user preferences
+     */
+    public Point computeTranslatedPoint(Point referencePoint, IFigure figure, boolean shiftWithGridValue) {
```

Comment:
change shiftWithGridValue to isFirstCreatedElement

Reply:
done